feat: add Lua APIs for prompt input masking.

cpu opened the pull request:
### Description

This commit adds three new Lua APIs: `prompt_mask.set`, `prompt_mask.get` and `prompt_mask.clear`.

The `prompt_mask.set` function accepts prompt data, and a table comprised of integer offsets to string content to insert into the prompt data at those locations pre-rendering. Mask content is not sent through to the MUD, or added to history data. Input wrapping is handled on the rendered prompt data as if the ANSI control characters were invisible. Reader mode avoids displaying any control characters in the prompt input buffer.

The function returns a bool indicating whether the mask was set. If false, the mask may have been invalid (out of range indexes, etc), or the prompt data may have already changed. Multiple calls to `set` will combine masks, with colliding keys having their values overwritten by the most recent `set` table.

When the input prompt buffer is changed we clear the prompt mask, ensuring that the two remain in sync. When a new prompt mask is set the prompt data is freshly rendered using the overlaid mask content.

The `prompt_mask.get` function returns the prompt mask table (if any). This is the combined view of all prior calls to `prompt_mask.set` and can be used to retrieve mask data.

The `prompt_mask.clear` function removes the prompt mask (if any) and redraws the input prompt data without masking.

The primary use-case for these APIs are to support scripts adding ANSI colour codes at predefined offsets to "highlight" words (e.g. based on a spellchecking result).

See `/help prompt_mask` for more information.

Resolves #723

cpu commented on src/model/prompt_mask.rs:
```diff
@@ -0,0 +1,55 @@
+            // NB: idx subtracted by one to account for Lua one-indexing.
+            let adjusted_idx = offset + (idx - 1) as usize;
+
+            // TODO(XXX): This is a HACK! Our mask/buf are falling out of sync somehow...
+            if adjusted_idx > masked_buf.len() {
+                return masked_buf.iter().collect::<String>();
+            }
```
I mention this in the WIP PR's TODO list. I've noticed if I enter some data with a prompt mask, then send the data, then reload the data from history, then press "down" to clear the input prompt, something weird happens:

* Pressing the down arrow ends up setting the prompt content to a new empty buffer.
* We clear the prompt mask when that happens
* But! Somehow `self.mask.iter` on L23 returns the old mask content, even though the mask has been cleared.
* This results in the `adjusted_idx` being outside the bounds of the `masked_buf` (which is empty), and a panic ensues.

I added this bounds check here as a temporary hack and it "solves" the problem. Maybe someone more familiar with the project can spot my error?

LiquidityC commented:
At first glance this looks great. I'm glad the idea with a prompt mask had the intended low impact and avoiding having to re-invent the command buffer to make this coloring work.
I'll take a peek and see if I can't see what's causing that buffer thing also. As for prompt wrapping. I don't remember how we draw it. But just skip all masks that are "before" the index where we render the prompt input from. That might be trickier then it sounds. As I said, don't really recall how it's implemented :smile:

cpu commented:
quick update: found the cause of the desync and implemented a fix locally. I'm looking at the prompt wrapping part next.

> skip all masks that are "before" the index where we render the prompt input from. 

makes sense. I think I see a way to make this work but it will require pushing the prompt mask down into the UI code that draws the prompt since that's where the wrapping happens.

On a related note: I'm thinking `ReaderScreen`'s impl of the `UserInterface` trait's `print_prompt_input` fn should ignore the masking entirely. Does that make sense to you or should I try to support that UI type as well?

LiquidityC commented:
> quick update: found the cause of the desync and implemented a fix locally. I'm looking at the prompt wrapping part next.
> 
> > skip all masks that are "before" the index where we render the prompt input from.
> 
> makes sense. I think I see a way to make this work but it will require pushing the prompt mask down into the UI code that draws the prompt since that's where the wrapping happens.

Yes. I hate that... If it starts to get messy I think we should take a step back and thing.
Perhaps we can create an iterator that only steps through printable chars perhaps? I know the prompt wrapping relies on the `chars()` iterators a lot.

> On a related note: I'm thinking `ReaderScreen`'s impl of the `UserInterface` trait's `print_prompt_input` fn should ignore the masking entirely. Does that make sense to you or should I try to support that UI type as well?

Yes. I see no situation where someone is using reader mode without actually relying on a screen reader. This is not the first case where we omit "colors" in `ReaderScreen`

cpu commented:
> Perhaps we can create an iterator that only steps through printable chars?

@LiquidityC This idea worked nicely, thanks! PTAL at the implementation. I'll work on adding docs + tests next. I think it will be ready to go after that.

<details>
<summary>I've been testing with this simple lua script to highlight words in the prompt area that are an even number of characters:</summary>

```lua
local red_on = "\x1b[31m";
local ansi_off = "\x1b[0m";

prompt.add_prompt_listener(function (buf)
    local words = {}
    local word_start = 1
    for i = 1, #buf do
        local c = buf:sub(i, i)
        local stop = c:find("[%s%p]")

        if stop ~= nil then
            local word = buf:sub(word_start, i - 1)
            words[word_start] = word
            word_start = i + 1
        elseif i == #buf then
            local word = buf:sub(word_start, i)
            words[word_start] = word
        end
    end

    local mask = {}
    for idx, word in pairs(words) do
        local even = #word % 2 == 0
        local off_idx = idx + #word
        if even then
            mask[idx] = red_on
            mask[off_idx] = ansi_off
        end
    end

    prompt_mask.set(buf, mask)
end)

```
</details>

<details>
<summary>Here's a video of the feature in action, showcasing history + prompt wrapping all working nicely. No more hacky workarounds:</summary>

[![asciicast](https://asciinema.org/a/v9QDne59s6AJUYcQgg8G9KFPs.svg)](https://asciinema.org/a/v9QDne59s6AJUYcQgg8G9KFPs)

</details>

cpu commented on Cargo.toml:
```diff
@@ -25,6 +25,7 @@ signal-hook = "0.3.14"
 mlua =  { version = "0.8.7", features = ["lua54", "send", "vendored"] }
 regex = "1.7.1"
 strip-ansi-escapes = "0.1.1"
+vte = "0.11.0"
```
quick note: I added this `vte` dependency to be able to parse out terminal control codes in `printable_chars.rs`. However, it is [already a transitive dependency via the `strip-ansi-escapes` crate](https://github.com/luser/strip-ansi-escapes/blob/a0a306ab6c5a76f269fea8659fdd736f1d2d56b9/Cargo.toml#L13-L14) so there's no increase in the net size of deps.

LiquidityC replied:
No problem

LiquidityC commented:
> > Perhaps we can create an iterator that only steps through printable chars?
> 
> @LiquidityC This idea worked nicely, thanks! PTAL at the implementation. I'll work on adding docs + tests next. I think it will be ready to go after that.

I'm really glad it worked. The impact on the original code is very small now which is always comforting. I'll take a closer look tomorrow. Nice work.

LiquidityC commented:
With some tweaking once this is in place it would be possible to write a plugin that highlights alias matches in the prompt input. So that you know if you are sending an alias before you're actually sending it. Perhaps not valuable. But it's cool.

cpu commented:
:wave: I force pushed some revisions. Some highlights:

* I added docs, under `/help prompt_mask`.
* I added support for calling `prompt_mask.set` multiple times. The mask tables are added together and colliding keys are replaced by the most recently provided value. 
* I added validation of the mask table. Providing a mask table with out of bound indexes or an index that doesn't fall at a clean character boundary will return false.
* There are unit tests now. A couple are a bit rough, so I might try to polish them up while this is under review but the core functionality is all tested.

<details>
<summary>Here's an updated test.lua that uses two masks:</summary>

```lua
-- convert buf to a table of words, indexed by their start location in buf.
function word_table(buf)
    local words = {}
    local word_start = 1
    for i = 1, #buf do
        local c = buf:sub(i, i)
        local stop = c:find("[%s%p]")

        if stop ~= nil then
            local word = buf:sub(word_start, i - 1)
            words[word_start] = word
            word_start = i + 1
        elseif i == #buf then
            local word = buf:sub(word_start, i)
            words[word_start] = word
        end
    end

    return words
end

-- set a prompt mask for buf where all even length words are highlighted red.
function even_word_mask(buf)
    local words = word_table(buf)

    local mask = {}
    for idx, word in pairs(words) do
        local even = #word % 2 == 0
        if even then
            local hi_on = BG_RED .. C_BWHITE
            local off_idx = idx + #word
            mask[idx] = hi_on
            mask[off_idx] = C_RESET
        end
    end

    prompt_mask.set(buf, mask)
end

-- set a prompt mask for buf where special words are highlighted green.
function special_word_mask(buf)
    local words = word_table(buf)

    local mask = {}
    for idx, word in pairs(words) do
        if word == "blightmud" then
            local hi_on = BG_GREEN .. C_BWHITE
            local off_idx = idx + #word
            mask[idx] = hi_on
            mask[off_idx] = C_RESET
        end
    end

    prompt_mask.set(buf, mask)
end

-- set up two prompt listeners for setting prompt masks.
prompt.add_prompt_listener(even_word_mask)
prompt.add_prompt_listener(special_word_mask)
```

</details>

<details>
<summary>Here's an updated demo video:</summary>

[![asciicast](https://asciinema.org/a/fvy2y6rnybVZI68IohGM7iURA.svg)](https://asciinema.org/a/fvy2y6rnybVZI68IohGM7iURA)
</details>

I'm going to leave this a draft PR until I've had a chance to do that last bit of polish but it's ready for review whenever you have time. Thanks!

LiquidityC commented:
I think this looks great. Nice work. I just have one request for added functionality. After that you can mark it as ready for review and we'll get this bad boy merged.

- Introduce a `prompt_mask.clear()` in the lua API
- Introduce a `prompt_mask.exists()` in the lua API

These two functions should make it at least possible for two separate "masking-plugins" to play nicer and not disrupt each other too much. If it becomes a major issue with colliding masks we can extend this to have "instanced" masks or something. But I don't see any point in making it that general yet.

cpu commented:
> Introduce a prompt_mask.clear() in the lua API

Done.

> Introduce a prompt_mask.exists() in the lua API

This one turns out to be a bit tricky. I can stash a `bool` for "mask exists" from the `set` API impl w/ `set_named_registry_value`, but when the mask is invalidated by user input there's no convenient way I can spot from the `command.rs` context to update that value to `false`, meaning the result will be stale if a prompt mask is set and then new input arrives and the prompt mask is cleared.

I also put some thought into whether we could make a nicer `prompt_mask.get()` API that would return the table. This ends up being tricky because we can't stash the user provided table in the `set()` call (e.g. like how `prompt.set` supports `prompt.get`) because it wouldn't be the full "combined" result. When we have the combined mask result in hand we're in the command context where there's no easy access to the Lua backend to update a named registry value.

I might be missing an easy solution. If not, I think we could get by without `exists/get`. WDYT?

cpu commented on src/ui/command.rs:
```diff
@@ -415,13 +439,13 @@ pub fn spawn_input_thread(session: Session) -> thread::JoinHandle<()> {
-                                    buffer.get_buffer(),
-                                    buffer.get_pos(),
-                                ))
-                                .unwrap();
```
Quick note on this change: while testing I noticed that key binding events (e.g. F1, etc) would send a `UserInputBuffer` event even though none of the events `check_command_binds` processes could provide new input. This meant that if you set a prompt mask from a key binding it would immediately be invalidated when the unchanged prompt input is sent as a `UserInputBuffer` event.

Moving where the event was emitted so it is conditional on `check_command_binds` returning `false` fixed my test script so that I could press "f4" and have the highlights appear, where previously pressing "f4" would briefly flash the mask before it was invalidated and cleared:


```lua
blight.bind("f3", function ()
    blight.output("clearing mask");
    prompt_mask.clear()
end)
blight.bind("f4", function ()
  local data = prompt.get()
  if #data == 0 then
    blight.output("no prompt data")
    return
  end
  local danger_mask = {[1] = BG_RED, [#data+1] = C_RESET};
  local res = prompt_mask.set(data, danger_mask);
  blight.output(string.format("masked: %s",res))
end)
```

LiquidityC replied:
Looks like you fixed a potential future bug in my eyes.

cpu commented:
> I might be missing an easy solution.

Oh, I think the solution is to add a method to `LuaScript` that does the needful with the Lua `state`. The command context has access to the session and the LuaScript inside. I'll give that a whirl after lunch. Sorry for the noise.

LiquidityC commented:
If there's no simple solution then skip it. Clear is the important one. The other is just QoL

cpu commented:
> If there's no simple solution then skip it. Clear is the important one. The other is just QoL

I just pushed a revision that adds `prompt_mask.get`. It wasn't too bad :+1:

LiquidityC commented:
> > Test & Build / Tests (ubuntu-latest, nightly) (pull_request) Failing after 4m
> 
> ```
> error: test failed, to rerun pass `--test communication_tests`
> 
> Caused by:
>   process didn't exit successfully: `/home/runner/work/Blightmud/Blightmud/target/debug/deps/communication_tests-345bac71090d49cc` (signal: 11, SIGSEGV: invalid memory reference)
> Error: Process completed with exit code 101.
> ```
> 
> I ran these tests locally ~2000 times and didn't catch this failure. Maybe this is similar flakyness to #726? thinking
> 
> _Edit: Further flake evidence: [it happened](https://github.com/Blightmud/Blightmud/actions/runs/4138419422/jobs/7154807856) against tip of dev testing a docs only PR_

Yes. I haven't been able to figure out what's causing it. But it's rare enough now that I can't really be bothered either. If you have any best guess to what the problem is then please share :smile:

LiquidityC commented on resources/help/prompt_mask.md:
```diff
@@ -0,0 +1,50 @@
+
+***prompt_mask.clear()***
+Clear the current prompt mask table (if any).
+
```
Nitpick: According to the "common pattern" there's usually a `##` separator between doc sections.

LiquidityC reviewed (changes requested):
All the code looks great. I just have one comment on one of the help files and it's a massiv nit-pick. Other then that everything looks great.
An added bonus is that there are almost no red sections (removals) in the diff which makes the chance for breaking existing functionality almost zero.

I think we cooked up a good solution here and you wrote good code. Thanks!

cpu commented:
> I think we cooked up a good solution here and you wrote good code.

Thanks for the help iterating on the design :-)  Just wanted to note that pushing the docs fix un-did the auto-merge.
